### DISCUSSION TITLE
Not powering off ports

### DISCUSSION THREAD

**Discussion by EdwinFairchild** [mvp/uhubctl#478] [Q&A]
The device I have is this [here](https://www.amazon.com/dp/B08YNPXPRW?psc=1&ref=ppx_yo2ov_dt_b_product_details)
and the output of `uhubctl -l 1-4 -a off -p 3`  is:

```
Current status for hub 2-4.4 [2109:0817 VIA Labs, Inc. USB3.0 Hub 000000000, USB 3.20, 4 ports, ppps]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 2-4 [2109:0817 VIA Labs, Inc. USB3.0 Hub 000000000, USB 3.20, 4 ports, ppps]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 0263 power 5gbps U3 enable connect [2109:0817 VIA Labs, Inc. USB3.0 Hub 000000000, USB 3.20, 4 ports, ppps]
Current status for hub 1-4.4 [2109:2817 VIA Labs, Inc. USB2.0 Hub 000000000, USB 2.10, 5 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
  Port 5: 0503 power highspeed enable connect [2109:8817 VIA Labs, Inc. USB Billboard Device 0000000000000001]
Current status for hub 1-4 [2109:2817 VIA Labs, Inc. USB2.0 Hub 000000000, USB 2.10, 5 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0000 off
  Port 4: 0503 power highspeed enable connect [2109:2817 VIA Labs, Inc. USB2.0 Hub 000000000, USB 2.10, 5 ports, ppps]
  Port 5: 0503 power highspeed enable connect [2109:8817 VIA Labs, Inc. USB Billboard Device 0000000000000001]
```

and here is the output of the `on` command:

```
Current status for hub 2-4.4 [2109:0817 VIA Labs, Inc. USB3.0 Hub 000000000, USB 3.20, 4 ports, ppps]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 2-4 [2109:0817 VIA Labs, Inc. USB3.0 Hub 000000000, USB 3.20, 4 ports, ppps]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 0263 power 5gbps U3 enable connect [2109:0817 VIA Labs, Inc. USB3.0 Hub 000000000, USB 3.20, 4 ports, ppps]
Current status for hub 1-4.4 [2109:2817 VIA Labs, Inc. USB2.0 Hub 000000000, USB 2.10, 5 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
  Port 5: 0503 power highspeed enable connect [2109:8817 VIA Labs, Inc. USB Billboard Device 0000000000000001]
Current status for hub 1-4 [2109:2817 VIA Labs, Inc. USB2.0 Hub 000000000, USB 2.10, 5 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0103 power enable connect [0403:6015 FTDI FT231X USB UART D30ALJQS]
  Port 4: 0503 power highspeed enable connect [2109:2817 VIA Labs, Inc. USB2.0 Hub 000000000, USB 2.10, 5 ports, ppps]
  Port 5: 0503 power highspeed enable connect [2109:8817 VIA Labs, Inc. USB Billboard Device 0000000000000001]

```
However the device does not lose power. Is this because it is externally powered? And are the devices on the list still in production I cant seem to find any of those amazon basics one.

**Comment by mvp** [comment]
This device is weird because it seems to have different number of ports for dual devices: USB3 `1-2` has 5 ports, while it's dual USB2 `2-4` has 4 ports. This confuses uhubctl dual matching logic. You should try `uhubctl -l 2-4 -p 3 -a 0` too, it will turn off dual port manually.

If that doesn't work, keep looking for devices in supported list - they shouldn't be too hard to find, especially on eBay.

**Reply by EdwinFairchild** [reply]
I tried that before and said no compatible devices found at 2-4

**Reply by mvp** [reply]
Which uhubctl version? This can only happen if you have older than 2.4.0. I recommend 2.5.0 or later.

**Reply by EdwinFairchild** [reply]
idk i just followed instruction in the readme, clone and build and install

**Reply by EdwinFairchild** [reply]
2.5.0-2-g19dfdad7

**Reply by mvp** [reply]
Which OS? If Linux, try running under sudo. If uhubctl shows hub 2-4, but trying to control it with sudo uhubctl says no compatible devices, that should not be possible and must be new bug.

**Reply by EdwinFairchild** [reply]
yeah im doing it all under sudo in Ubuntu

**Reply by EdwinFairchild** [reply]
![image](https://user-images.githubusercontent.com/62710807/209766243-71df6be8-26cc-4a47-8d2a-4c6533585a9a.png)

**Reply by mvp** [reply]
Thanks. One last question: can you run in the same console
```
sudo uhubctl --version
```
Just trying to make sure you don't have system installed old uhubctl anywhere.

**Reply by EdwinFairchild** [reply]
![image](https://user-images.githubusercontent.com/62710807/209766890-dcfabb18-1b9e-405c-9e52-da1702de8311.png)

**Reply by mvp** [reply]
This must be new bug. I can't reproduce it :-(

If you have any gdb skills, you can try debugging it. Basically if uhubctl shows hub in the list, it must be to control it too...

**Reply by mvp** [reply]
Quick cheat sheet for gdb debugging if you want to try it:
```
make
gdb ./uhubctl
l    # print source code, enter to keep printing
b <line>   # set breakpoint
c    # continue execution until breakpoint
bt   # backtrace - print current call stack if at breakpoint or crashed
```

**Reply by EdwinFairchild** [reply]
i use gdb on the daily but on mcu targets , ill give it a shot

**Reply by mvp** [reply]
Few more
```
n    # execute next line, don't step in
s    # execute next line, step into subroutine if any
r     # restart from scratch
p <variable>   # print variable value
q    # quit
```

**Reply by EdwinFairchild** [reply]
so not sure how gdb works on a binary like this.... ?  how to give uhubctl the troubling command and where t place the breakpoint

**Reply by EdwinFairchild** [reply]
![image](https://user-images.githubusercontent.com/62710807/209768365-f42fdbd8-04ee-4be1-8448-04969de982e0.png)

**Reply by mvp** [reply]
You need run it on binary which was created by make. If uhubctl.c exists in the same directory, it should work automatically

**Reply by mvp** [reply]
To pass arguments, use `set args -l 2-4...`

**Reply by mvp** [reply]
One more tip: debugging and sudo don't mix well. Instead, I recommend setting up udev permissions which make it work without sudo - there is readme FAQ for this.

**Reply by EdwinFairchild** [reply]
im battling with udev rules right now
```
SUBSYSTEM=="usb", DRIVER=="hub", MODE="0666"
# Linux 6.0 or later (its ok to have this block present for older Linux kernels):
SUBSYSTEM=="usb", DRIVER=="hub", \
  RUN="/bin/sh -c \"chmod -f 666 $sys$devpath/*-port*/disable || true\""
```
```
SUBSYSTEM=="usb", DRIVER=="hub", MODE="0666"
# Linux 6.0 or later (its ok to have this block present for older Linux kernels):
SUBSYSTEM=="usb", DRIVER=="hub", \
  RUN="/bin/sh -c \"chmod -f 666 $sys$devpath/*-port*/disable || true\""
  ```
also tried with not filter

**Reply by mvp** [reply]
There are reports that `DRIVER="usb"` works better, can you give it a try?

**Reply by mvp** [reply]
Basically one line:
```
SUBSYSTEM=="usb", DRIVER=="usb", MODE="0666"
```

**Reply by EdwinFairchild** [reply]
Running gdb with sudo because  i live dangerously

```
eddie@DeusXMachina:~/playground/uhubctl$ sudo gdb --args ./uhubctl -l 2-4 -p 3 -a 0
[sudo] password for eddie: 
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./uhubctl...
(gdb) run
Starting program: /home/eddie/playground/uhubctl/uhubctl -l 2-4 -p 3 -a 0
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff7d29640 (LWP 3866)]
No compatible devices detected at location 2-4!
Run with -h to get usage info.
[Thread 0x7ffff7d29640 (LWP 3866) exited]
[Inferior 1 (process 3863) exited with code 01]
(gdb) 

```

**Reply by EdwinFairchild** [reply]
yeah that one liner worked

**Reply by mvp** [reply]
That makes no sense 😹. If `uhubctl ` shows device 2-4, it must be able to control it too. I guess try to backtrack it why it cannot find it while it can list it.

**Reply by EdwinFairchild** [reply]
![image](https://user-images.githubusercontent.com/62710807/209770840-3cb2b138-50c0-4355-ba3b-c289734417eb.png)

**Reply by mvp** [reply]
So `usb_find_hubs()` returned 0.
Perhaps trace it step by step inside why

**Reply by EdwinFairchild** [reply]
ok ill give this a whirl tomorrow because right  now im so sleepy i might pass out  at my desk . I post my findings

**Reply by EdwinFairchild** [reply]
What I can gather from stepping through the code  is that [this line](https://github.com/mvp/uhubctl/blob/19dfdad7f6c00fa9c5e3b7d926589782300fc20b/uhubctl.c#L880) sets `info.actionable` to `0` 
which ultimately means [this line](https://github.com/mvp/uhubctl/blob/19dfdad7f6c00fa9c5e3b7d926589782300fc20b/uhubctl.c#L1008) is always true and thus `hub_phys_count` is still at `0` and returns `0`

**Reply by mvp** [reply]
Hmm. Can you try adding -e, e.g.
```
uhubctl -l 2-4 -p 3 -a 0 -e
```
Basically what's going on is that uhubctl tries to find matching USB2 hub, but because port count doesn't match (4 vs 5), it gets confused and doesn't count 2-4 as physical hub. Normally, any USB3 hub will always have matching USB2 compatibility dual partner.

**Reply by mvp** [reply]
And -e completely disables any dual matching logic, so hub 2-4 will be treated as physical hub of its own.

**Reply by EdwinFairchild** [reply]
It loses data connection but not power

**Reply by EdwinFairchild** [reply]
```
Check for VBUS (voltage) off support: plug a phone, USB light or USB fan into USB port of your hub. Try using uhubctl to turn power off on that port, and check that phone stops charging, USB light stops shining or USB fan stops spinning. If VBUS doesn't turn off, your hub manufacturer did not include circuitry to actually cut power off. Such hub would still work to cut off USB data connection, but it cannot turn off power, and we do not consider this supported device
```

I believe this is the case

**Reply by mvp** [reply]
That's unfortunate. I guess this hub doesn't properly support power switching :-(

**Reply by EdwinFairchild** [reply]
i think ill just pick up one of the supported devices , because it will stop me from splicing into the power line and adding a mosfet just to programmatically control power, this tool seems heaven sent if I can get it to work. 
Do you have any links to hubs that you yourself know are working with this?

**Reply by mvp** [reply]
I am trying really hard to only add hubs that actually work. Most of them are reported by other people - but reporters are asked to do vbus test before it is added to the table. I personally own and attest definitely working following hardware: D-Link DUB-H7 (silver), Rosonway RSH-A16, UUgear MEGA4, and Microchip USB2517.

**Reply by EdwinFairchild** [reply]
jeez that Rosonway RSH-A16 is beastly , ill give that one a try. thanks

**Reply by mvp** [reply]
Warning about RSH-A16: I have one with Realtek chipset which works great. However I was told that most currently sold samples are using Via chipset and don't work. It is impossible to tell them apart just looking at retail box. You might want to contact Rosonway directly.

DUB-H7 silver can be purchased on eBay for $9 now. Avoid black one - most likely won't work.

**Reply by EdwinFairchild** [reply]
dam, about to cancel that order then

**Comment by yongyuandouneng** [comment]


**Reply by EdwinFairchild** [reply]
this is probably the first time i see spam on github

**Comment by maximvelichko** [comment]
I have few of those 2109:2817/2109:0817 (USB 2.0/USB 3.0)

It works as expected for USB 2.0 devices, and errors out with `No compatible devices detected at location 1-2!` for USB 3.0 devices.
